Data/Functor/Foldable/TH.hs

```diff
@@ -0,0 +1,359 @@
+{-# LANGUAGE Rank2Types #-}
+module Data.Functor.Foldable.TH
+  ( makeBaseFunctor
+  , makeBaseFunctorWith
+  , BaseRules (..)
+  , baseRules
+  ) where
+
+import Control.Applicative as A
+import Data.Traversable as T
+import Data.Bifunctor (first)
+import Data.Functor.Identity
+import Language.Haskell.TH
+import Language.Haskell.TH.Syntax (mkNameG_tc, mkNameG_v)
+import Data.Char (GeneralCategory (..), generalCategory)
+import Data.Orphans ()
+#ifndef CURRENT_PACKAGE_KEY
+import Data.Version (showVersion)
+import Paths_recursion_schemes (version)
+#endif
+
+-- | Build base functor with a sensible default configuration.
+--
+-- /e.g./
+--
+-- @
+-- data Expr a
+--     = Lit a
+--     | Add (Expr a) (Expr a)
+--     | Expr a :* [Expr a]
+--   deriving (Show)
+-- @
+--
+-- will create
+--
+-- @
+-- data ExprF a x
+--     = LitF a
+--     | AddF x x
+--     | x :*$ [x]
+--   deriving ('Functor', 'Foldable', 'Traversable')
+--
+-- type instance 'Base' (Expr a) = ExprF a
+--
+-- instance 'Recursive' (Expr a) where
+--     'project' (Lit x)   = LitF x
+--     'project' (Add x y) = AddF x y
+--     'project' (x :* y)  = x :*$ y
+--
+-- instance 'Corecursive' (Expr a) where
+--     'embed' (LitF x)   = Lit x
+--     'embed' (AddF x y) = Add x y
+--     'embed' (x :*$ y)  = x :*$ y
+-- @
+--
+-- @
+-- 'makeBaseFunctor' = 'makeBaseFunctorWith' 'baseRules'
+-- @
+--
+-- /Notes:/
+--
+-- 'makeBaseFunctor' works properly only with ADTs.
+-- Existentials and GADTs aren't supported,
+-- as we don't try to de better than
+-- <https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#deriving-functor-instances GHC's DeriveFunctor>.
+--
+makeBaseFunctor :: Name -> DecsQ
+makeBaseFunctor = makeBaseFunctorWith baseRules
+
+-- | Build base functor with a custom configuration.
+makeBaseFunctorWith :: BaseRules -> Name -> DecsQ
+makeBaseFunctorWith rules name = reify name >>= f
+  where
+    f (TyConI dec) = makePrimForDec rules dec
+    f _            = fail "makeBaseFunctor: Expected type constructor name"
+
+-- | Rules of renaming data names
+data BaseRules = BaseRules
+    { _baseRulesType  :: Name -> Name
+    , _baseRulesCon   :: Name -> Name
+    , _baseRulesField :: Name -> Name
+    }
```

Comment:
* A small blurb describing what in particular each of these fields are modifying would be helpful.
* I noticed each of these functions are of type `Name -> Name`. Do we really need the full power of `Name` here? Or if these are only for renaming generated things, could we get away with just `String -> String` (like what [`aeson` does](http://hackage.haskell.org/package/aeson-1.0.2.1/docs/Data-Aeson-TH.html#v:fieldLabelModifier))?
* Do these need to begin with underscores? All of `BaseRules`' fields are exported, after all.